arbor/partition_load_balance.cpp

```diff
@@ -1,234 +1,219 @@
     }
-
-    for (unsigned i = 0; i < super_cells.size(); i++) {
-        auto kind = rec.get_cell_kind(super_cells[i].front());
-        for (auto gid: super_cells[i]) {
-            if (rec.get_cell_kind(gid) != kind) {
-                throw gj_kind_mismatch(gid, super_cells[i].front());
-            }
-            local_gids.push_back(gid);
-        }
-        kind_lists[kind].push_back({i, true});
+    if (hint.prefer_gpu && !hint.gpu_group_size) {
+        throw arbor_exception(arb::util::pprintf("unable to perform load balancing because {} has invalid suggested gpu_cell_group size of {}",
+                                                 kind, hint.gpu_group_size));
     }
+    if (hint.prefer_gpu && has_gpu) return std::make_pair(backend_kind::gpu, hint.gpu_group_size);
+    return std::make_pair(backend_kind::multicore, hint.cpu_group_size);
+}
 
-
-    // Create a flat vector of the cell kinds present on this node,
-    // partitioned such that kinds for which GPU implementation are
-    // listed before the others. This is a very primitive attempt at
-    // scheduling; the cell_groups that run on the GPU will be executed
-    // before other cell_groups, which is likely to be more efficient.
-    //
-    // TODO: This creates an dependency between the load balancer and
-    // the threading internals. We need support for setting the priority
-    // of cell group updates according to rules such as the back end on
-    // which the cell group is running.
-
-    auto has_gpu_backend = [&ctx](cell_kind c) {
-        return cell_kind_supported(c, backend_kind::gpu, *ctx);
-    };
-
-    std::vector<cell_kind> kinds;
-    for (auto l: kind_lists) {
-        kinds.push_back(cell_kind(l.first));
+struct group_parameters {
+    cell_kind kind;
+    backend_kind backend;
+    size_t size;
+};
+
+// Create a flat vector of the cell kinds present on this node, sorted such that
+// kinds for which GPU implementation are listed before the others. This is a
+// very primitive attempt at scheduling; the cell_groups that run on the GPU
+// will be executed before other cell_groups, which is likely to be more
+// efficient.
+//
+// TODO: This creates an dependency between the load balancer and the threading
+// internals. We need support for setting the priority of cell group updates
+// according to rules such as the back end on which the cell group is running.
+auto build_group_parameters(context ctx,
+                            const partition_hint_map& hint_map,
+                            const std::unordered_map<cell_kind, std::vector<cell_gid_type>>& kind_lists) {
+    std::vector<group_parameters> res;
+    for (const auto& [kind, _gids]: kind_lists) {
+        const auto& [backend, group_size] = get_backend(ctx, kind, hint_map);
+        res.push_back({kind, backend, group_size});
     }
-    std::partition(kinds.begin(), kinds.end(), has_gpu_backend);
+    util::sort_by(res, [](const auto& p) { return p.kind; });
```

Comment:
I guess this is where the order in the enum matters?

Reply:
Yes!